snuba/query/parser/expressions.py

```diff
@@ -101,65 +123,51 @@ def visit_low_pri_arithmetic(
         self,
         node: Node,
         visited_children: Tuple[
-            Any, Expression, Any, Union[Node, Tuple[LowPriOperator, Expression]]
+            Any, Expression, Any, Union[Node, LowPriTuple, List[LowPriTuple]]
         ],
     ) -> Expression:
         _, term, _, exp = visited_children
 
-        if exp == "":
+        if isinstance(exp, Node):
             return term
 
+        if isinstance(exp, LowPriTuple):
+            if exp.op == LowPriOperator.PLUS:
+                return plus(term, exp.arithm)
+            elif exp.op == LowPriOperator.MINUS:
+                return minus(term, exp.arithm)
+
         for elem in exp:
-            if isinstance(elem, list):
-                # If there remains Expressions
-                # in List form that need to be
-                # put together and bubbled up
-                if elem[0] == LowPriOperator.PLUS:
-                    term = plus(term, elem[1])
-                elif elem[0] == LowPriOperator.MINUS:
-                    term = minus(term, elem[1])
-            else:
-                # This branch is hit when
-                # there is only one operator left
-                # to perform arithmetic between two
-                # Expressions that have been bubbled up
-                if exp[0] == LowPriOperator.PLUS:
-                    return plus(term, exp[1])
-                elif exp[0] == LowPriOperator.MINUS:
-                    return minus(term, exp[1])
+            if elem.op == LowPriOperator.PLUS:
+                term = plus(term, elem.arithm)
+            elif elem.op == LowPriOperator.MINUS:
+                term = minus(term, elem.arithm)
 
         return term
 
     def visit_high_pri_arithmetic(
         self,
         node: Node,
         visited_children: Tuple[
-            Any, Expression, Any, Union[Node, Tuple[HighPriOperator, Expression]]
+            Any, Expression, Any, Union[Node, HighPriTuple, List[HighPriTuple]]
         ],
     ) -> Expression:
         _, term, _, exp = visited_children
 
-        if exp == "":
+        if isinstance(exp, Node):
             return term
 
+        if isinstance(exp, HighPriTuple):
+            if exp.op == HighPriOperator.MULTIPLY:
+                return multiply(term, exp.arithm)
+            elif exp.op == HighPriOperator.DIVIDE:
+                return divide(term, exp.arithm)
```

Comment:
Same, you can avoid the duplication

Reply:
There's also a fair bit of duplication between `visit_low_pri_arithmetic` and `visit_high_pri_arithmetic`. Another idea could be to store a mapping between the operators and functions at the top of the file then have these two methods share the same implementation since they would be pretty much identical then.

Reply:
Hi, I'm not quite sure I follow just "mapping." Is there a specific implementation/procedure you are referring to? I don't think I've ever tried something like this (so I don't know how to approach this).

Reply:
Oh, I meant having something along the lines of

```
def get_arithmetic_function(operator) -> FunctionCall:
    return {
        HighPriOperator.MULTIPLY: multiply,
        HighPriOpereator.DIVIDE: divide,
        LowPriOperator.ADD: add,
        LowPriOperator.SUBTRACT: subtract,
    }[operator]
```

Then later both visit_low_pri_arithmetic and visit_high_pri_arithmetic would share a common implementation that does something like
```
return get_arithmetic_function(exp.op)(term, exp.arithm)
```

Does that idea make sense? Just another option, up to you whether you think it's useful or not.

Reply:
Yes, that idea makes sense. I'm adding it in!